### ISSUE TITLE
Some issues with the inference results

### ISSUE THREAD

**Issue by y200504040u** [ultralytics/ultralytics#15743]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Bug

I trained segmentation based on yolov8n-seg.yaml on my custom dataset. 
The input size(==640) of the network is equal to the original image size.
I got some strange mask results during inference. 
The ltrb of the mask looks like this, and only the ltrb of the mask has this issue, while other edges of mask are normal.

I checked my ground truth and confirmed that they are correct. 

![20240822-140957](https://github.com/user-attachments/assets/e3de264c-4813-40b5-9b51-addd552587e6)
![20240822-141006](https://github.com/user-attachments/assets/b1cfecd6-5778-471a-a8df-65fc0976b4d1)
![20240822-141009](https://github.com/user-attachments/assets/09317f45-1d54-4638-93af-a7d4f530e6e4)

**The ltrb of output are convex outward after training and inference based on the highest version 8.2.79. See the above.** 

**When I lower the version to 8.1.46 and perform training followed by inference, the ltrb of output are concave inward, like below.**

![20240822-173948](https://github.com/user-attachments/assets/8b2f121e-35ff-4919-b1a0-014974e06f64)
![20240822-173943](https://github.com/user-attachments/assets/3c48fbd1-d0a9-4a65-befe-13f68169255a)

The same data and config are used in both cases.

It appears there might be an issue with the bbox branch in terms of ltrb positioning either during the training phase or the inference phase, as I do not think different versions should yield different results.

### Environment

YOLO 8.2.79
Python 3.8.9
torch 1.8.1
CUDA 11.1

### Minimal Reproducible Example

```python
import os
import cv2
import torch
import numpy as np
from ultralytics import YOLO

path_root = "./imgs"
color = (255, 0, 0)
height, width = 480, 640
thre_conf = 0.7
list_color = [(255,0,0), (0,255,0), (0,0,255), (255,255,0), (255,0,255), (0,255,255), (0,153,255), (153,0,255)]

class SegYoloV8:
    def __init__(self, path_ckpt):
        self.model = YOLO(path_ckpt)
        self.list_rst = []
        self.names = None

    def get_mask_yolo(self, img_bgr):
        self.list_rst = []
        with torch.no_grad():
            results = self.model(img_bgr)
        result = results[0].cpu()
        if result is None or result.masks is None:
            print("Segment Failed.")
            self.mask = None
            self.bbox = None
        else:
            self.names = result.names
            list_masks = []
            list_bboxes = []
            for m, b in zip(result.masks.data, result.boxes.data):
                if b[4] > thre_conf:
                    list_masks.append(m)
                    list_bboxes.append(b)
            for mask, bbox in zip(list_masks, list_bboxes):
                dict_rst = {"mask": mask.numpy(), "bbox": bbox[:4].numpy(), "score": float(bbox[4]), "cls": int(bbox[5])}
                self.list_rst.append(dict_rst)
            print(self.names)

        torch.cuda.empty_cache()

if __name__ == '__main__':
    list_imgs_bgr = []
    segger = SegYoloV8("./best.pt")
    total = 0
    for dir_path, dir_names, file_names in os.walk(path_root):
        for name in file_names:
            if name.endswith(".jpg") and "rgb" in dir_path:
                if total % 10 == 0:
                    path_bgr = os.path.join(dir_path, name)
                    list_imgs_bgr.append(path_bgr)
                total += 1

    for path_bgr in list_imgs_bgr:
        img_bgr = cv2.imread(path_bgr)
        segger.get_mask_yolo(img_bgr)
        if len(segger.list_rst) > 0:
            for rst in segger.list_rst:
                mask = rst['mask']
                bbox = rst['bbox'].astype(np.uint16)
                cls = rst['cls']
                name = segger.names[cls]
            
                mask_ch3 = cv2.cvtColor(mask, cv2.COLOR_GRAY2BGR)
                mask_alpha = mask_ch3 * list_color[cls]
                plus = np.where(mask_ch3, mask_alpha, img_bgr)
                img_bgr = cv2.addWeighted(img_bgr.astype(np.float32), 0.6, plus.astype(np.float32), 0.4, 0).astype(np.uint8)
                cv2.putText(img_bgr, name, bbox[:2], cv2.FONT_HERSHEY_COMPLEX_SMALL, 0.8, (0,0,255))
        else:
            print("Failed to seg.", path_bgr)

        cv2.imshow("dst", img_bgr)
        key = cv2.waitKey(0)
        if key == ord('q') or key == 27:
            break
```

**Comment by glenn-jocher** [comment]
@y200504040u thank you for reporting this issue. Please ensure you are using the latest version of the Ultralytics package. If the problem persists, could you provide a minimal reproducible example and any additional details to help us investigate further?

**Comment by Y-T-G** [comment]
Try this

https://github.com/ultralytics/ultralytics/issues/14365#issuecomment-2225105580

**Comment by y200504040u** [comment]
> Try this
> 
> [#14365 (comment)](https://github.com/ultralytics/ultralytics/issues/14365#issuecomment-2225105580)

Thanks for your reply. I will try it.

**Comment by y200504040u** [comment]
> Thanks for your reply.
> I am using the latest version of the Ultralytics package and then lower the version to 8.1.46 to train and inference for debug.
> [https://github.com/ultralytics/ultralytics/issues/15747](url)
> A minimal reproducible example as mentioned in the above link.

Thanks for your reply.
I am using the latest version of the Ultralytics package and then lower the version to 8.1.46 to train and inference for debug.
[https://github.com/ultralytics/ultralytics/issues/15747](https://github.com/ultralytics/ultralytics/issues/url)
A minimal reproducible example as mentioned in the above link.

**Comment by y200504040u** [comment]
> @y200504040u thank you for reporting this issue. Please ensure you are using the latest version of the Ultralytics package. If the problem persists, could you provide a minimal reproducible example and any additional details to help us investigate further?

Thanks for your reply.
I am using the latest version of the Ultralytics package and then lower the version to 8.1.46 to train and inference for debug.
[https://github.com/ultralytics/ultralytics/issues/15747](https://github.com/ultralytics/ultralytics/issues/url)
A minimal reproducible example as mentioned in the above link.

**Comment by glenn-jocher** [comment]
@y200504040u thank you for confirming. Please provide the minimal reproducible example as mentioned in the link so we can investigate further.

**Comment by y200504040u** [comment]
> @y200504040u thank you for confirming. Please provide the minimal reproducible example as mentioned in the link so we can investigate further.

I just updated the link address. Thanks a lot.
[[https://github.com/ultralytics/ultralytics/issues/15747#issue-2480358640](https://github.com/ultralytics/ultralytics/issues/url)](url)

**Comment by Y-T-G** [comment]
Can you close that issue and paste the code here?

There's no need for two different issues.

**Comment by Y-T-G** [comment]
Does this issue occur with `result.plot()`?

**Comment by y200504040u** [comment]
> Does this issue occur with `result.plot()`?
After referring to your method, I've basically solved this problem. Thank you.
[[#14365]https://github.com/ultralytics/ultralytics/issues/14365#issuecomment-2306329782](url)

**Comment by y200504040u** [comment]
> Does this issue occur with `result.plot()`?

I did not use `result.plot()`. I wrote a piece of code to draw the results.

**Comment by glenn-jocher** [comment]
Thank you for the update. Could you please try using `result.plot()` to see if the issue persists? This will help us determine if the problem lies within the custom drawing code or the model's output.
